repo/content/index_blob_manager.go

```diff
@@ -0,0 +1,456 @@
+			return nil, errors.Wrap(err, "unable to read compaction log entry %q")
+		}
+
+		le.age = latestServerBlobTime.Sub(cb.Timestamp)
+
+		results[cb.BlobID] = le
+	}
+
+	return results, nil
+}
+
+func (m *indexBlobManagerImpl) deleteOldBlobs(ctx context.Context, latestBlob blob.Metadata) error {
+	allCompactionLogBlobs, err := m.listCache.listBlobs(ctx, compactionLogBlobPrefix)
+	if err != nil {
+		return errors.Wrap(err, "error listing compaction log blobs")
+	}
+
+	// look for server-assigned timestamp of the compaction log entry we just wrote as a reference.
+	// we're assuming server-generated timestamps are somewhat reasonable and time is moving
+	compactionLogServerTimeCutoff := latestBlob.Timestamp.Add(-m.maxEventualConsistencySettleTime)
+	compactionBlobs := blobsOlderThan(allCompactionLogBlobs, compactionLogServerTimeCutoff)
+
+	log(ctx).Debugf("fetching %v/%v compaction logs older than %v", len(compactionBlobs), len(allCompactionLogBlobs), compactionLogServerTimeCutoff)
+
+	compactionBlobEntries, err := m.getCompactionLogEntries(ctx, compactionBlobs)
+	if err != nil {
+		return errors.Wrap(err, "unable to get compaction log entries")
+	}
+
+	allCleanupBlobs, err := m.listCache.listBlobs(ctx, cleanupBlobPrefix)
+	if err != nil {
+		return errors.Wrap(err, "error listing cleanup blobs")
+	}
+
+	cleanupEntries, err := m.getCleanupEntries(ctx, latestBlob.Timestamp, allCleanupBlobs)
+	if err != nil {
+		return errors.Wrap(err, "error loading cleanup blobs")
+	}
+
+	indexBlobsToDelete := m.findIndexBlobsToDelete(ctx, latestBlob.Timestamp, compactionBlobEntries)
+	compactionLogBlobsToDelayCleanup := m.findCompactionLogBlobsToDelayCleanup(ctx, compactionBlobs)
+
+	if err := m.delayCleanupBlobs(ctx, compactionLogBlobsToDelayCleanup); err != nil {
+		return errors.Wrap(err, "unable to schedule delayed cleanup of blobs")
+	}
+
+	compactionLogBlobsToDelete, cleanupBlobsToDelete := m.findBlobsToDelete(cleanupEntries)
+
+	// note that we must always delete index blobs first before compaction logs
+	// otherwise we may inadvertedly resurrect an index blob that should have been removed.
+	if err := m.deleteBlobsFromStorageAndCache(ctx, indexBlobsToDelete); err != nil {
+		return errors.Wrap(err, "unable to delete compaction logs")
+	}
+
+	if err := m.deleteBlobsFromStorageAndCache(ctx, compactionLogBlobsToDelete); err != nil {
+		return errors.Wrap(err, "unable to delete compaction logs")
+	}
+
+	if err := m.deleteBlobsFromStorageAndCache(ctx, cleanupBlobsToDelete); err != nil {
+		return errors.Wrap(err, "unable to delete cleanup blobs")
+	}
+
+	m.flushCache()
+
+	return nil
+}
+
+func (m *indexBlobManagerImpl) findIndexBlobsToDelete(ctx context.Context, latestServerBlobTime time.Time, entries map[blob.ID]*compactionLogEntry) []blob.ID {
+	tmp := map[blob.ID]bool{}
+
+	for _, cl := range entries {
+		for _, b := range cl.InputMetadata {
+			if age := latestServerBlobTime.Sub(b.Timestamp); age < m.maxEventualConsistencySettleTime {
```

Comment:
The age comparison here should be done using the creation time of the compaction log and not of the individual index blobs. Essentially, the decision should be made based on **when the compaction was performed**, and thus the new output index blobs and compaction blobs were created, and not when the input blobs were originally created.

Pushed a sample change to address this at https://github.com/julio-lopez/kopia/pull/6/files

Reply:
per offline discussion, not sure if this matters since we're filtering compaction logs to include only the "settled" ones, but fixed anyway.